Крестики Нолики

GameShowLDA opened the pull request:
1.  Состоит из главного класса, который вызывает класс структуры игры
2.  Проверка комбинаций
3.  Расцветка хода
4.  Проверка корректности данных

GameShowLDA committed:
Крестики Нолики

1.  Состоит из главного класса, который вызывает класс структуры игры
2.  Проверка комбинаций
3.  Расцветка хода
4.  Проверка корректности данных

Dezodemius reviewed (commented):
Тебе нужно добавить в корневую папку репозиторий файл `.gitignore`. В нём нужно прописать:
```
*.suo
*.user
.vs/
[Bb]in/
[Oo]bj/
``` 

В идеале, конечно, найти в интернете .gitignore для VisualStudio и просто скопировать его содержимое.
Потом закоммить это всё отдельным коммитом

GameShowLDA committed:
Merge branch 'main' into TikTacToe-main

GameShowLDA committed:
Добавление файлов в gitignore

GameShowLDA committed:
.gitignore

Исправление файла gitignore (В частности название файла)

GameShowLDA committed:
Удаление gitignore

GameShowLDA committed:
Удаление других проектов

Dezodemius commented on TicTacToe/Program.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace TicTacToe
+{
+    class Program
```
Здесь и в других классах нужно оформить по соглашению

Dezodemius commented on TicTacToe/App.config:
Может, удалить его? Раз изменений нет. Да и не нужен он

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+    /// </summary>
+    class Game
+    {
+        #region [ ПЕРЕМЕННЫЕ ]
```
Регионы немного не так пишутся. В соглашении есть как надо оформить. Если коротко, то нужно писать просто 
`#region Поля и свойства`. В других местах аналогично

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+        bool boolFinishGame = false;
+        // Переменная для начала новой игры
+        bool boolStartGame = true;
+        // Игровое поле
```
Документация не так оформляется. Нужно через xml-доку

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+        /// <summary>
+        /// Выбор хода
+        /// </summary>
+        private void ChoosingAMove()
```
Артикли не нужно в названиях писать

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+            boolActiveMove = true;
+        }
+        /// <summary>
+        /// Выбор хода
```
В конце доки нужно ещё и точку ставить

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+                try
+                {
+                    int x = (str[0] - '0') - 1;
+                    int y = (str[2] - '0') - 1;
```
Вот такое обращение через индексы, тем более которые не по порядку, очень хрупкое. Я бы предложил несколько вариков:
1. Сделать одномерный массив и запрашивать одно число от 1 до 9
2. Не дать пользователю вводить пробел, а просто два числа рядом
3. Считывать каждое число по очереди и обособленно проверять их

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+
+                    if (charArrMoves[x, y] == ' ')
+                    {
+                        if (boolActiveMove) charArrMoves[x, y] = 'X';
```
'X' в константу или enum нужно вывести

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+                    Console.WriteLine("Координата введена неверно");
+                }
+
+                if (CheckFinishGame() == 'X') { Console.ForegroundColor = ConsoleColor.Green; Console.WriteLine("Победа крестиков!"); boolFinishGame = true; }
```
Зачем же в одну строку? :/

Dezodemius replied:
Ниже аналогично нужно сделать - не в одну строку

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -0,0 +1,213 @@
+        /// <summary>
+        /// Перезапуск игры
+        /// </summary>
+        private void RetutnGame()
```
Дока имени не соответствует

Dezodemius commented on Task1/App.config:
А точно задания по практике нужно удалять? Если слить эту ветку так, то у  тебя удалится всё в мастере

GameShowLDA committed:
Исправление недочётов

Изменение xml документации
Продуманное название переменных
Изменение проверки хода

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -11,37 +11,85 @@ namespace TicTacToe
-        // Победные координаты
-        char charWin = ' ';
+        #region Константы
+        /// <summary>
```
Объявление региона/закрытие региона должно отделяться пустой строкой от членов класса

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -11,37 +11,85 @@ namespace TicTacToe
     /// </summary>
     class Game
```
Модификатор доступа нужон

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -11,37 +11,85 @@ namespace TicTacToe
     /// </summary>
```
В конце доки должна быть точка

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -11,37 +11,85 @@ namespace TicTacToe
+        /// <summary>
+        ///  X - Ход со стороны крестиков
+        /// </summary>
+        private const char crossMove = 'X';
```
Константы называются с большой буквы

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -11,37 +11,85 @@ namespace TicTacToe
+        /// true - ход крестиком
+        /// false - ход ноликом
+        /// </summary>
+        private bool boolActiveMove = true;
```
Какое-то странное решение для смены хода

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -11,37 +11,85 @@ namespace TicTacToe
+        /// <summary>
+        /// Игровое поле
+        /// </summary>
+        private string[] stringArrPlayingField = new string[9];
```
А зачем два массива? Почему нельзя всё в одном делать?

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -54,7 +102,7 @@ private void DefaultSettings()
                 for (int j = 0; j < 3; j++)
                 {
-                    charArrMoves[i, j] = ' ';
+                    arrMoves[i, j] = ' ';
```
Не нужно писать сокращения в названии переменных

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -63,26 +111,25 @@ private void DefaultSettings()
         /// Выбор хода
         /// </summary>
-        private void ChoosingAMove()
+        private void ChoosingMove()
```
`ChooseMove`

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -63,26 +111,25 @@ private void DefaultSettings()
-        private void ChoosingAMove()
+        private void ChoosingMove()
         {
             while (!boolFinishGame)
```
Не стоит писать тип переменной в её названии. Нужно давать такие названия, чтобы было понятно, что оно bool

GameShowLDA replied:
Это привычка с работы)

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -63,26 +111,25 @@ private void DefaultSettings()
-                    int x = (str[0] - '0') - 1;
-                    int y = (str[2] - '0') - 1;
+                    Console.Write("Выберите строку (от 1 до 3): ");
+                    int x = Convert.ToInt32(Console.ReadLine()) - 1;
```
А если неудачная конвертация будет? Всё упадёт

GameShowLDA replied:
Всё нормально будет, там конструкция try catch

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -97,11 +144,26 @@ private void ChoosingAMove()
-                if (CheckFinishGame() == 'X') { Console.ForegroundColor = ConsoleColor.Green; Console.WriteLine("Победа крестиков!"); boolFinishGame = true; }
-                else if (CheckFinishGame() == 'O') { Console.ForegroundColor = ConsoleColor.Green; Console.WriteLine("Победа ноликов!"); boolFinishGame = true; }
-                else if (CheckFinishGame() == 'D') { Console.ForegroundColor = ConsoleColor.Green; Console.WriteLine("Победила дружба!"); boolFinishGame = true; }
+                if (CheckFinishGame() == crossMove)
```
Можно ведь один раз выполнить CheckFinishGame, а потом сравнивать

Dezodemius commented on TicTacToe/Game.cs:
```diff
@@ -115,49 +177,97 @@ private void ChoosingAMove()
-            // Проверка диагоналей                                                                                     
-            else if (charArrMoves[0, 0] != ' ' && charArrMoves[0, 0] == charArrMoves[1, 1] && charArrMoves[1, 1] == charArrMoves[2, 2]) { charWin = charArrMoves[0, 0]; charArrMoves[0, 0] = charArrMoves[1, 1] = charArrMoves[2, 2] = 'W'; PlayingField(); return charWin; }
-            else if (charArrMoves[0, 2] != ' ' && charArrMoves[0, 2] == charArrMoves[1, 1] && charArrMoves[1, 1] == charArrMoves[2, 0]) { charWin = charArrMoves[0, 2]; charArrMoves[0, 2] = charArrMoves[1, 1] = charArrMoves[2, 0] = 'W'; PlayingField(); return charWin; }
+            if (arrMoves[0, 0] != ' ' && arrMoves[0, 0] == arrMoves[0, 1] && arrMoves[0, 0] == arrMoves[0, 2])
```
Будто бы, в конце строки очень много пробелов. Их нужно убрать

Dezodemius replied:
В других местах тоже есть

Dezodemius reviewed (changes requested):
Нужно поправить

GameShowLDA committed:
Внос попровок

Новый алгоритм вывода поля в консоль
Правка "Внешнего вида" кода
